src/api/metadata.coffee

```diff
@@ -0,0 +1,161 @@
+        try
+          uidObj = getUniqueIdentifierForCollection key, doc
+          uid = uidObj[Object.keys(uidObj)[0]]
+          result = yield collections[key].find(uidObj).exec()
+          if result and result.length > 0 and result[0]._id
+            status = 'Conflict'
+          else
+            doc = new collections[key] doc
+            error = doc.validateSync()
+            return throw new Error "Validation failed: #{error}" if error
+            status = 'Valid'
+          
+          logger.info "User #{this.authenticated.email} successfully inserted #{key} with unique identifier #{uid}"
+          returnObject.successes.push buildResponseObject key, doc, status, 'Ok', uid
+          
+        catch e
+          logger.error "Failed to insert #{key} with unique identifier #{uid}. #{e.message}"
+          returnObject.errors.push buildResponseObject key, doc, 'Invalid', e.message, uid
+        
+    this.body = returnObject
+    this.status = 201
+  catch e
+    this.body = e.message
+    utils.logAndSetResponse this, 500, "Could not fetch all users via the API #{e}", 'error'
+
+
+# API endpoint that updates or inserts metadata from import
+exports.upsertMetadata = () ->
+  # Test if the user is authorised
+  if not authorisation.inGroup 'admin', this.authenticated
+    utils.logAndSetResponse this, 403, "User #{this.authenticated.email} is not an admin, API access to getUsers denied.", 'info'
+    return
+  
+  try
+    returnObject = {"errors":[], "successes":[]}
+    insertObject = this.request.body
+    
+    for key of insertObject
+      return utils.logAndSetResponse this, 400, "Could not fetch all users via the API: 'Invalid Request Body'", 'error' if key not of collections
+      insertDocuments = insertObject[key]
+      for doc in insertDocuments
+        try
+          uidObj = getUniqueIdentifierForCollection key, doc
+          uid = uidObj[Object.keys(uidObj)[0]]
+          result = yield collections[key].find(uidObj).exec()
+          if result and result.length > 0 and result[0]._id
+            delete doc._id if doc._id
+            yield collections[key].findByIdAndUpdate(result[0]._id, doc).exec()
+            status = 'Successfully Updated'
+          else
+            doc = new collections[key] doc
+            result = yield Q.ninvoke doc, 'save'
+            status = 'Successfully Inserted'
+          
+          logger.info "User #{this.authenticated.email} successfully inserted #{key} with unique identifier #{uid}"
+          returnObject.successes.push buildResponseObject key, doc, status, 'Ok', uid
+          
+        catch e
+          logger.error "Failed to insert #{key} with unique identifier #{uid}. #{e.message}"
+          returnObject.errors.push buildResponseObject key, doc, 'Error', e.message, uid
+        
+    this.body = returnObject
+    this.status = 201
+  catch e
+    this.body = e.message
+    utils.logAndSetResponse this, 500, "Could not fetch all users via the API #{e}", 'error'
```

Comment:
The code between this function and the previous one is basically identical. It would be better to combine these into one function so that the code isn't repeated. Remember the DRY principle :)